Assets/HoloToolkit-Preview/QRTracker/Scripts/QRCodesManager.cs

```diff
@@ -0,0 +1,160 @@
+﻿using System;
+using System.Collections;
+
+using System.Collections.Generic;
+
+using UnityEngine;
+
+using QRCodesTrackerPlugin;
+namespace HoloToolkit.Unity.QRTracking
+{
+    public static class QRCodeEventArgs
+    {
+        public static QRCodeEventArgs<TData> Create<TData>(TData data)
+        {
+            return new QRCodeEventArgs<TData>(data);
+        }
+    }
+
+    [Serializable]
+    public class QRCodeEventArgs<TData> : EventArgs
+    {
+        public TData Data { get; private set; }
+
+        public QRCodeEventArgs(TData data)
+        {
+            Data = data;
+        }
+    }
+
+    public class QRCodesManager : Singleton<QRCodesManager>
+    {
+        [Tooltip("Determines if the QR codes scanner should be automatically started.")]
+        public bool AutoStartQRTracking = true;
+
+        public bool IsTrackerRunning { get; private set; }
+        public QRCodesTrackerPlugin.QRTrackerStartResult StartResult { get; private set; }
+
+
+        public event EventHandler<QRCodeEventArgs<QRCodesTrackerPlugin.QRCode>> QRCodeAdded;
+        public event EventHandler<QRCodeEventArgs<QRCodesTrackerPlugin.QRCode>> QRCodeUpdated;
+        public event EventHandler<QRCodeEventArgs<QRCodesTrackerPlugin.QRCode>> QRCodeRemoved;
+
+        private System.Collections.Generic.SortedDictionary<System.Guid, QRCodesTrackerPlugin.QRCode> qrCodesList = new SortedDictionary<System.Guid, QRCodesTrackerPlugin.QRCode>();
+
+        private QRTracker qrTracker;
+
+        public System.Guid GetIdForQRCode(string qrCodeData)
+        {
+            lock (qrCodesList)
```

Comment:
This is more of a meta discussion, but locking an object you're modifying can be tricky. Usually you want another lock object to be more explicit. I don't see any offending code here, but if we did something like
`lock (qrCodesList) { qrCodesList = new ... }` we'd hit problems

Reply:
Agreed, I think the async utilities doesn't play nicely with the lock as well bc of the way [Unity's synchronization context is implemented](https://github.com/svermeulen/Unity3dAsyncAwaitUtil/issues/10) as well. (But we won't need to worry about that until we bring this over to MRTK and if we make this method asynchronous).

Reply:
🤮

Reply:
The only downside to using the concurrent queue is that it technically has different behavior. Here, and in the issue you linked, the queue gets drained all at once before allowing more work to be added. We could get in a state where the thread locks waiting for the queue to drain while continually adding work. 

I wonder if a CopyTo would be the best approach if we wanted to snapshot the queue per execution call...

Queue also has a [SyncRoot](https://docs.microsoft.com/en-us/dotnet/api/system.collections.queue.syncroot?view=netframework-4.7.2), but I'm not sure that's really any better. It at least would work if someone used the Queue as a [Synchronized](https://docs.microsoft.com/en-us/dotnet/api/system.collections.queue.synchronized?view=netframework-4.7.2)

Reply:
>We could get in a state where the thread locks waiting for the queue to drain while continually adding work.

That's exactly how I ended up fixing it in my own synchronization context implementation.

Reply:
```

        internal static void Post(Action task)
        {
            lock (_Actions)
            {
                _Actions.Enqueue(task);
            }
        }


        private void Update()
        {
            int actionCount;

            lock (_Actions)
            {
                actionCount = _Actions.Count;
            }

            for (int i = 0; i < actionCount; i++)
            {
                Action next;

                lock (_Actions)
                {
                    next = _Actions.Dequeue();
                }

                next();
            }
        }
```